### ISSUE TITLE
Mypy can't figure out that value is not-None when indexing by variable

### ISSUE THREAD

**Issue by sofiaml** [python/mypy#11534]
The following code:
```
def read(test: Mapping[str, Optional[str]], key: str) -> str:
    if test[key] is not None:
        return test[key]
    return "default"
```
gives the error `test.py:6: error: Incompatible return value type (got "Optional[str]", expected "str")`, despite having checked that `test[key]` is not None. I believe this mypy error is a false positive.

(For reference, something like this has no error:
```
def read(test: Mapping[str, Optional[str]], key: str) -> str:
    if test["a"] is not None:
        return test["a"]
    return "default"
```
)

**Comment by erictraut** [comment]
This isn't specific to a "not None" type guard. More generally, type narrowing of subscript expressions isn't supported when the subscript expression is dynamic. As you've noted, it is supported when the subscript expression is a string or numeric literal value. 

Type narrowing for dynamic subscript expressions is very difficult to implement because it requires the type checker to ensure that the _value_ of the expression found within the type guard test evaluates the same _value_ as the expression found within the guarded block. That's difficult to do when the expression is dynamic. I'm not aware of any type checkers or compilers (in any language) that implement such support.

A type-friendly workaround is to use an assignment expression, like this:

```python
def read(test: Mapping[str, Optional[str]], key: str) -> str:
    if (x := test[key]) is not None:
        return x
    return "default"
```

**Comment by sofiaml** [comment]
Thank you very much for the explanation and the workaround!
